packages/utils/src/instrument.ts

```diff
@@ -357,179 +357,228 @@ function instrumentHistory(): void {
-    // a 'keypress' will reset the keypress debounce so that a new
-    // set of keypresses can be recorded
-    keypressTimeout = undefined;
     // It's possible this handler might trigger multiple times for the same
-    // event (e.g. event propagation through node ancestors). Ignore if we've
-    // already captured the event.
+    // event (e.g. event propagation through node ancestors).
+    // Ignore if we've already captured that event.
     if (!event || lastCapturedEvent === event) {
       return;
     }
 
-    lastCapturedEvent = event;
+    // We always want to skip _some_ events.
+    if (shouldSkipDOMEvent(event)) {
+      return;
+    }
+
+    const name = event.type === 'keypress' ? 'input' : event.type;
 
-    if (debounceTimer) {
-      clearTimeout(debounceTimer);
+    // If there is no debounce timer, it means that we can safely capture the new event and store it for future comparisons.
+    if (debounceTimerID === undefined) {
+      handler({
+        event: event,
+        name,
+        global: globalListener,
+      });
+      lastCapturedEvent = event;
     }
 
-    if (debounce) {
-      debounceTimer = setTimeout(() => {
-        handler({ event, name });
+    // If there is a debounce awaiting, see if the new event is different enough to treat it as a unique one.
+    // If that's the case, emit the instrumentation event and store locally newly captured DOM event.
+    if (debounceTimerID !== undefined && shouldShortcircuitPreviousDebounce(lastCapturedEvent, event)) {
+      handler({
+        event: event,
+        name,
+        global: globalListener,
       });
-    } else {
-      handler({ event, name });
+      lastCapturedEvent = event;
     }
+
+    // Start a new debounce timer that will prevent us from capturing multiple events that should be grouped together.
+    clearTimeout(debounceTimerID);
+    debounceTimerID = global.setTimeout(() => {
+      debounceTimerID = undefined;
+    }, debounceDuration);
   };
 }
 
-/**
- * Wraps addEventListener to capture keypress UI events
- * @param handler function that will be triggered
- * @returns wrapped keypress events handler
- * @hidden
- */
-function keypressEventHandler(handler: Function): (event: Event) => void {
-  // TODO: if somehow user switches keypress target before
-  //       debounce timeout is triggered, we will only capture
-  //       a single breadcrumb from the FIRST target (acceptable?)
-  return (event: Event): void => {
-    let target;
+type AddEventListener = (
+  type: string,
+  listener: EventListenerOrEventListenerObject,
+  options?: boolean | AddEventListenerOptions,
+) => void;
+type RemoveEventListener = (
+  type: string,
+  listener: EventListenerOrEventListenerObject,
+  options?: boolean | EventListenerOptions,
+) => void;
+
+type InstrumentedElement = Element & {
+  __sentry_instrumentation_handlers__?: {
+    [key in 'click' | 'keypress']?: {
+      handler?: Function;
+      rc: number;
```

Comment:
What is `rc`? Can we either include a comment explaining or give it a more descriptive name?